### PR TITLE
Stake

### PR DESCRIPTION
**Pull request by Purva-Chaudhari**

### CODE REVIEW

File: src/RWAStaking.sol [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "./stRWA.sol";
+import "./GOON.sol";
+import "./NEST.sol";
+import "./interfaces/IDateTime.sol";
+import "./interfaces/ICheckin.sol";
+import "./CheckInStorage.sol";
+
+contract RWAStaking is UUPSUpgradeable, AccessControlUpgradeable {
+    //using CheckInStorage for CheckInStorage.Storage;
```
Comment (eyqs) [inline src/RWAStaking.sol:15]: In the next PR: `using RWAStorage for RWAStorage.Storage`

File: src/RWAStaking.sol [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        require(_goonUsdToken.transfer(msg.sender, goonUsdAmount), "goonUSD transfer failed");
+    }
+
+    function claim() public {
+    accumulateRewards(msg.sender);
+
+    uint256 rewards = unclaimedRewards[msg.sender];
+    
+    if (rewards > 0) {
+        uint256 pRewards = (rewards * 2)/1000;
```
Comment (eyqs) [inline src/RWAStaking.sol:77]: In the next PR:

- Initializer: `rs.goonDivisor = 500`
- Setter: `function setGoonDivisor(uint256 divisor) adminOnly`
- Calculation: `uint256 goonRewards = rewards / rs.goonDivisor`

File: src/RWAStaking.sol [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        require(_goonToken.transfer(msg.sender, pRewards), "$Goon transfer failed");
+
+        uint256 nestRewards = rewards * 2;
+        require(_nestToken.transfer(msg.sender, nestRewards), "$NEST transfer failed");
+
+        rewards = 1 + ((rewards * 2)/100 - 1) / 1e18;
+        checkIn.updateStakePoints(msg.sender, rewards);
```
Comment (eyqs) [inline src/RWAStaking.sol:84]: Same for NEST and miles in the next PR:

- Initializer: `rs.nestMultiplier = 2`
- Setter: `function setNestMultiplier(uint256 multiplier) adminOnly`
- Calculation: `uint256 nestRewards = rewards * rs.nestMultiplier`

File: src/RWAStaking.sol [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    } else {
+        rewards = 0;
+    }
+
+    unclaimedRewards[msg.sender] = 0;
+}
+
+
+    function setAPR(uint256 APR) external onlyRole(DEFAULT_ADMIN_ROLE)  {
+        _APR = APR;
```
Comment (eyqs) [inline src/RWAStaking.sol:94]: In the next PR:

- Initializer: `rs.APR = 40`

File: src/RWAStaking.sol [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        accumulateRewards(msg.sender);
+
+        uint256 goonUsdAmount = _stRwaToken.balanceOf(msg.sender);
+        _stRwaToken.burn(msg.sender, amount);  
+
+        require(_goonUsdToken.transfer(msg.sender, goonUsdAmount), "goonUSD transfer failed");
+    }
+
+    function claim() public {
+    accumulateRewards(msg.sender);
```
Comment (eyqs) [inline src/RWAStaking.sol:72]: Fix indentation

File: lib/openzeppelin-contracts
Comment (eyqs) [inline lib/openzeppelin-contracts:1]: Run `forge install` again - there shouldn't be any changes

File: src/CheckIn.sol
Comment (eyqs) [inline src/CheckIn.sol:1]: Remove all of these changes

File: src/CheckInStorage.sol [hunk 6]
```diff
@@ -62,6 +62,7 @@ library CheckInStorage {
         uint256 basePoints;
         uint256 faucetPoints;
         uint256 swapPoints;
+        uint256 stakePoints;
```
Comment (eyqs) [inline src/CheckInStorage.sol:65]: Remove

File: src/interfaces/ICheckin.sol [hunk 7]
```diff
@@ -1,9 +1,11 @@
...
 pragma solidity ^0.8.14;
+import "../CheckInStorage.sol";
 
 interface ICheckIn {
     function getReRolls(address user) external view returns (uint256);
     function incrementPoints(address user, uint8 tier) external;
     function incrementFaucetPoints(address user, string memory token) external;
-    function incrementTaskPoints(address user, uint256 task) external;
+    function incrementTaskPoints(address user, CheckInStorage.Task task) external;
+    function updateStakePoints(address user, uint256 points) external;
```
Comment (eyqs) [inline src/interfaces/ICheckin.sol:10]: Remove `updateStakePoints`, keep the rest of the changes

File: src/interfaces/ICheckin.sol [hunk 8]
```diff
@@ -1,9 +1,11 @@
 // SPDX-License-Identifier: UNLICENSED
 pragma solidity ^0.8.14;
```
Comment (eyqs) [inline src/interfaces/ICheckin.sol:2]: Can you increase this to 0.8.25?

File: src/RWAStaking.sol [hunk 9]
```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.14;
```
Comment (eyqs) [inline src/RWAStaking.sol:2]: Increase to 0.8.25

File: src/stRWA.sol [hunk 10]
```diff
@@ -1,86 +1,210 @@
...
-    ERC20Upgradeable,
-    ERC20BurnableUpgradeable,
-    ERC20PausableUpgradeable,
-    UUPSUpgradeable
-{
-    string private _name;
-    string private _symbol;
-
-    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+pragma solidity ^0.8.0;
```
Comment (eyqs) [inline src/stRWA.sol:2]: Increase to 0.8.25

File: script/DeployTestnetContracts.s.sol [hunk 11]
```diff
@@ -28,14 +28,12 @@ contract DeployScript is Script {
         console.log("DateTime deployed to:", address(dateTime));
 
         address checkInProxy = Upgrades.deployUUPSProxy(
-            "CheckIn.sol",
-            abi.encodeCall(CheckIn.initialize, (msg.sender, address(dateTime), address(dateTime)))
+            "CheckIn.sol", abi.encodeCall(CheckIn.initialize, (msg.sender, address(dateTime), address(dateTime)))
```
[the remark this answers is not in the document]
Comment (Purva-Chaudhari) [reply script/DeployTestnetContracts.s.sol:31]: yes Linter did it

File: lib/openzeppelin-foundry-upgrades
[the remark this answers is not in the document]
Comment (Purva-Chaudhari) [reply lib/openzeppelin-foundry-upgrades:1]: I removed cached before commiting. It wont affect local builds

**Review by eyqs — APPROVED** [review]
LGTM

**Pull request merged**